### ISSUE TITLE
Containers causing unexpected physics in child Node2D bodies

### ISSUE THREAD

**Issue by jeroenheijmans** [godotengine/godot#74940]
### Godot version

4.0.stable.official [92bee43ad]

### System information

Windows 10, OpenGL API 3.3.0 NVIDIA 528.49, NVIDIA GeForce GTX 1080

### Issue description

I'm trying to use GUI controls to create a layout for my game, with in the center of the screen a fixed rectangle that shows the Node2D and physics stuff of my game. To arrange things in this window I used `HBoxContainer`, but this affects my physics in very unpredictable manners.

I'm a beginner at Godot, using version 4 (I've never used v3.x), so I'm not sure if this behavior is a _bug_ (although it seems so to me), or a usability or documentation issue (because it's for sure _unexpected_ and _unpredictable_).

### Steps to reproduce

I've added a reproducible scenario in [a GitHub repository](https://github.com/jeroenheijmans/repro-godot-bounce-bug), but it's quite minimal to construct it. The basic details:

1. New Godot 4.0 project with default project settings
2. Create node structure like this:

    ```none
    Node
       \Control
           \HBoxContainer
               \ColorRect (Custom minimum size 200 x 100 px)
             \PanelContainer (Custom minimum size 400 x 600 px)
                 \RigidBody2D (with PhysicsMaterial, Bounce = 0.5)
                     \CollisionShape2D (circle)
                 \CharacterBody2D (set Transform.Scale.x = -1 but leave y at 1)
                     \CollisionPolygon2D (sloped platform, see screenshot)
    ```

3. Turn on "Visible Collision Shapes" in the debug window (since there are no sprites to see the effects)
4. Run the scene

- ❌ Result: the ball ends up _inside_ the polygon, slowly sliding down
- ✅ Expected result: the ball bouncing a bit off the platform

Interesting variants:

- ✅ Remove the `ColorRect` => expected result now happens
- ❌ Change `HBoxContainer` to `VFlowContainer`, `GridContainer`, or probably any other `Container` => the unexpected result happens
- ❌ Skip step 3 (so leave the `Scale.x` to `1`) => unexpectedly the ball bounces off _really_ fast

Video of the ❌ unexpected result with the base repro:

https://user-images.githubusercontent.com/1590536/225328868-fae41a2c-fff4-4888-9b3a-d1f2f3949c34.mp4

Video of the ❌ unexpected _variant_ with non-flipped CharacterBody2D polygon:

https://user-images.githubusercontent.com/1590536/225336577-0f4b7942-8ca5-4c28-96e1-6d535c5be60a.mp4

Video of the expected result (after removing `ColorRect` from the aforementioned tree):

https://user-images.githubusercontent.com/1590536/225329993-e2e41067-0dd3-4394-a29b-7db113f66cfc.mp4

### Minimal reproduction project

Clone and run the following Godot 4 project to reproduce the issue: https://github.com/jeroenheijmans/repro-godot-bounce-bug

**Comment by jeroenheijmans** [comment]
> I think this is not a bug, but user error.

Yikes, that -especially the second half- ("user error") is very harsh (and hard) to read after spending a full day of reading documentation, debugging, and writing a detailed report 😢

In the intro text of my message above I also already admit it might not be a _bug_. But "user error" doesn't feel appropriate to me either? If "_the two don't mix_" then:

- possibly it shouldn't be _possible_ to actually do mix them?
- probably one or more of the relevant documentation sections should mention this? (I read most of the [Introduction to Godot](https://docs.godotengine.org/en/stable/getting_started/introduction/introduction_to_godot.html), the [Nodes and Scenes](https://docs.godotengine.org/en/stable/getting_started/step_by_step/nodes_and_scenes.html) Step-by-step, the top level [User interface (UI)](https://docs.godotengine.org/en/stable/tutorials/ui/index.html) page)
- it could be nice if there was a warning in the Scene Tree (similar to the "Node configuration warning" you get when adding a `StaticBody2D` without any children)?

In short, even though I understand that things might not be "broken" (that it's not a bug), perhaps there are still ways to prevent users from running into this situation?

**Comment by YuriSizov** [comment]
@jeroenheijmans Godot allows you to put any node as a child of any other node. This is by design, and there are also cases where some nodes which don't mix would be totally fine together (especially if you don't expect them to affect each other). There is no limitation on the engine side, except for a wrong expectation. So adding a configuration warning, for example, wouldn't be helpful to a lot of users, unless someone has the same expectation as you did here.

You generally shouldn't expect nodes from different systems to affect each other when composed in a tree. Most of the time they just won't do anything to each other. Sometimes they can break things that propagate through chains of similar nodes. I don't believe the documentation pages that you've listed suggest that mixing GUI nodes with 2D nodes to achieve what you were set to achieve is possible. You arrived there on your own, and that's totally okay 🙂  You were learning the tool and trying to solve the problem at hand, and had a wrong assumption, which happens sometimes. I recommend joining one of the communities to get feedback and support when looking for a solution and ending up with something that is not working.

**Comment by jeroenheijmans** [comment]
Alright, understood.
